### PR TITLE
fix(Scripts/ZulGurub): Rewrite High Priest Thekal

### PR DESCRIPTION
**Pull request by Nyeriah**
## Changes Proposed:
-  Move everything to TaskScheduler
-  Reset the cats once boss resets
- Phase 2 starts once everybody dies
- Adds ressurrect within 10 seconds if not everyone is dead
- Fixed some scuffed stuff with thekal attacking people while lying on the floor

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/11507

## Tests Performed:

- Tested in-game
- 

## How to Test the Changes:

1. Pull thekal & test his abilities... pretty much everything needs testing
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Gultask** [comment]
I'll update this in case I find more stuff.

- [ ] I can manage to pull one tiger less. GUIDs are: Tigers (49311, 49312) Zath (49313) Lor'Khan (49314) Thekal (49310)
Videos show that they aren't supposed to be linked as to pull everyone immediately, though. ([source](https://www.youtube.com/watch?v=-vRbHAJ1r3Y))

WoWScrnShot_052422_142758

- [ ] When a reset happens, it seems that any and all tigers within a radius of Thekal are resurrected, not necessarily the 2 near him.

https://user-images.githubusercontent.com/100873791/170099878-5903db21-47e3-4825-ad7e-64368853d6e7.mp4

- [ ] Zealot Lor'Khan isn't healing herself, and is healing Thekal even if he is at full health. ([source](https://youtu.be/EokJmnTM3Bw?t=33))

- [ ] Dispel Magic is missing from the code, and it should be castable by Lor'Khan. Bloodlust is also attributed to her instead of Thekal. ([source](https://wowpedia.fandom.com/wiki/High_Priest_Thekal))

**Comment by Nyeriah** [comment]
I have made the cat issue better but it's not fixed yet, I'll address it in another PR. It needs to be fixed through formations but formations needs a few changes to work with this correctly as I see it, so I'll postpone it.

I made the zealot heal whoever has the least health, let me know if it feels better.

Transition issue fixed, it was some extra check I had added to prevent them from reseting or hitting unintended stuff.

**Comment by Gultask** [comment]
>I made the zealot heal whoever has the least health, let me know if it feels better.

I don't know if it's blizzlike but it feels much better 👍 
Although she still isn't healing herself.

- [ ] Lor'Khan still casts spells while downed. I saw her casting Lightning Shield and Greater Heal.

- [ ] Thekal carries an axe in Tiger form. (Same thing happened to Mar'li 😄)

WoWScrnShot_052522_191228

- [ ] Tiger form is too large. I like it though.

- [ ] The Zealots and Thekal should perform death animations and not the /lie & /sleep one

- [ ] Thekal should yell his RAGE line during phase 2 transition, not at pull

- [ ] There should be a Hakkar line played when a reset happens: "Do not be lured into the mortals' trap. Bathe in my power! Be renewed and destroy the heretics!" ([source](https://youtu.be/VFraeEvi2BQ?t=20))

- [ ] Transition not blizzlike. Once everyone is dead, Thekal revives almost immediately, has a pause, performs his line, then transforms.

Sources: 
https://youtu.be/UHJJCcgtIhw?t=142 
https://youtu.be/ENBkZvaSYu8?t=74
https://youtu.be/ENBkZvaSYu8?t=76 I don't know what happened there with the early revive with no health.

Nefertum didn't say that I shouldn't nitpick so I'll try to find bugs until it's more blizzlike than retail. 😁

**Comment by temperrr** [comment]
[ ] There is no dead animation on the zealots they just stand upright staring creepingly at you
Other issues seems fixed from last comment

**Comment by Nyeriah** [comment]
Hm your compile must’ve broke, that definitely doesn’t happen

**Review by acidmanifesto — APPROVED** [review]

File: src/server/scripts/EasternKingdoms/ZulGurub/boss_thekal.cpp [hunk 1]
```diff
@@ -82,170 +66,240 @@ class boss_thekal : public CreatureScript
...
 
         void UpdateAI(uint32 diff) override
         {
-            if (!UpdateVictim())
+            if (me->GetReactState() != REACT_PASSIVE && !UpdateVictim())
                 return;
 
-            events.Update(diff);
+            _scheduler.Update(diff,
+                std::bind(&BossAI::DoMeleeAttackIfReady, this));
```
[the remark this answers is not in the document]
Comment (Nyeriah) [reply src/server/scripts/EasternKingdoms/ZulGurub/boss_thekal.cpp:227]: TrinityCore standards/personal preference

**Comment by Nyeriah** [comment]
> Maybe move the scripts to new script system?

After Silker's mandokir pr is merged then, since it's introduced there and I'd rather not create conflicts

**Pull request merged**